uswds/uswds
# USWDS - In-page navigation: Exclude hidden headers

# Summary
**Updated JavaScript to exclude hidden headers from the in-page navigation link list.** Any header with a style of `display:none` or `visibility:hidden` will now be excluded from the list of links in the component.

> **Note**
> This PR does not exclude headers hidden with `usa-sr-only` or a style of `opacity: 0`. The conversation surrounding this decision is in [this comment](https://github.com/uswds/uswds/pull/5393#pullrequestreview-1543854409), [this comment](https://github.com/uswds/uswds/pull/5393#issuecomment-1666252566), and [this comment](https://github.com/uswds/uswds/pull/5393#issuecomment-1667872517).

> **Warning**
> Both PR [#5387](https://github.com/uswds/uswds/pull/5387) and this PR edit the same files and functions. We should expect merge conflicts.

## Breaking change

This is not a breaking change.

## Related issue

Closes #5392

## Related pull requests
[Changelog PR](https://github.com/uswds/uswds-site/pull/2197)

## Preview link

Preview link: [In-page navigation](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-in-page-nav-hidden-header/?path=/story/components-in-page-navigation--default)

## Problem statement

The in-page navigation component pulls hidden headers from the main content region. These should be excluded from the component link list. 

## Solution

The `getSectionHeadings` function now filters through the array of section headings and removes any headers that are styled with `display:none` or `visibility:hidden`.

## Testing and review
To test in Storybook:
Open the [Hidden Headers test page](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-in-page-nav-hidden-header/?path=/story/components-in-page-navigation--test-hidden-header) and confirm that only the visible headers are included in the in-page navigation link list. 

To test in a local project:
1. Install this branch
2. Add component code for the [in-page page navigation](https://designsystem.digital.gov/components/in-page-navigation/#in-page-navigation-code) to a test page.
3. Add h2 or h3 headers that are hidden with `display: none` or `visibility:hidden` styles to the `main` content region.
5. Check that only the visible headers are included in the in-page nav link list

Additionally: 
- Confirm that this accounts for all expected hidden header scenarios
- Confirm the code follows USWDS patterns
- Confirm the test Storybook pages are useful, make sense, and are structured appropriately
- Confirm no visual or behavioral regressions
